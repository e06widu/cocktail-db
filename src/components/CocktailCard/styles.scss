@import "../../style/variables";
.card-container {
    border: 1px solid $white;
    padding: 20px;
    text-align: center;
    .thumb-image {
        width: 100%;
    }
    .c-name,
    .c-category {
        letter-spacing: 0.2px;
    }
    .c-name {
        margin-bottom: 10px;
    }
    .c-category {
        margin-top: 10px;
    }
    Button {
        margin-top: 10px;
    }
    .btn {
        .icon {
            margin-right: 5px;
        }
    }
    .already-in {
        margin-top: 20px;
    }
}

@media screen and (max-width: 280px) {
    h1 {
        margin-top: 0;
    }
    .c-name,
    .c-category {
        font-size: 14px;
    }
}

@media screen and (max-width: 480px) {}