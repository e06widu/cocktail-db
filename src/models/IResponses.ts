import { ICocktail } from "./ICocktail";

export type ISearchCocktailResponse = {
    drinks: ICocktail[];
};

export type IRandomCocktailResponse = {
    drinks: ICocktail[];
};
export type IMultipleRandomCocktailResponse = {
    drinks: ICocktail[];
};

