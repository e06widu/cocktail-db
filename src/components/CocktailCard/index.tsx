import React, { useEffect, useState } from 'react'
import { ICocktail } from '../../models/ICocktail'

import './styles.scss'
import { useQuery, useQueryClient } from '@tanstack/react-query'
import Button from '../Button'
import Toast from '../Toast'

interface ICocktailCardProps {
    cocktail: ICocktail
    mode: 'home' | 'favorites' | 'search'
    handleRemoveFromCartToast?: (cocktail: ICocktail) => void
}

const CocktailCard: React.FC<ICocktailCardProps> = ({ cocktail, mode, handleRemoveFromCartToast }) => {
    const { data: favorites } = useQuery<ICocktail[]>(['cocktail', 'favorites']);
    const queryClient = useQueryClient();
    const [showToast, setShowToast] = useState<boolean>(false);

    const [isAlreadyInFavorite, setIsAlreadyInFavorite] = useState<boolean>(false)

    useEffect(() => {
        if (favorites)
            setIsAlreadyInFavorite(favorites.some(item => item.idDrink === cocktail.idDrink));
    }, [cocktail.idDrink, favorites]);

    useEffect(() => {
        if (showToast) {
            const timer = setTimeout(() => {
                setShowToast(false);
            }, 3000);

            return () => {
                clearTimeout(timer);
            };
        }
    }, [showToast]);

    const handleAddToCart = () => {
        queryClient.setQueryData<ICocktail[]>(['cocktail', 'favorites'], (oldData) =>
            [...(oldData || []), cocktail]);
        setShowToast(true);
    };

    const handleRemoveFromCart = () => {
        queryClient.setQueryData<ICocktail[]>(['cocktail', 'favorites'], (oldData) =>
            oldData?.filter((item) => item.idDrink !== cocktail.idDrink)
        );
        if(handleRemoveFromCartToast){
            void handleRemoveFromCartToast(cocktail);
        }
    };

    return (
        <div className='card-container'>
            <div>{cocktail.strDrink}</div>
            {<img className='thumb-image' src={cocktail.strDrinkThumb}
                alt={cocktail.strDrink} loading="lazy" />}

            {mode === 'home' ? <div>{cocktail.strCategory}</div> : null}
            {mode === 'search' ?
                isAlreadyInFavorite ? <div>Already added</div> : <Button disabled={isAlreadyInFavorite} onClick={() => void handleAddToCart()}>Add</Button>
                : null}
            {mode === 'favorites' ? <Button onClick={() => void handleRemoveFromCart()}>Remove</Button> : null}
            {showToast && <Toast message={`Cocktail ${cocktail.strDrink} Added`} />}
        </div>
    )
}

export default CocktailCard
