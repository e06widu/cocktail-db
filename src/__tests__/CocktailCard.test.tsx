import { render, screen, fireEvent } from '@testing-library/react';
import CocktailCard from '../components/CocktailCard';
import { ICocktail } from '../models/ICocktail';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

describe('CocktailCard', () => {
    const mockCocktail = {
        idDrink: '1',
        strDrink: 'Mojito',
        strDrinkThumb: 'mojito.jpg',
        strCategory: 'Cocktail',
    } as ICocktail;

    test('renders cocktail name and image', () => {
        render(
            <QueryClientProvider client={new QueryClient()}>
                <CocktailCard cocktail={mockCocktail} mode="home" />
            </QueryClientProvider>
        );
        expect(screen.getByText('Mojito')).toBeDefined();
        const image = screen.getByAltText('Mojito');
        expect(image.getAttribute('src')).toBe('mojito.jpg');
    });


    test('handles add to cart button click', () => {
        const setQueryDataMock = jest.fn();
        jest.spyOn(QueryClient.prototype, 'setQueryData').mockImplementation(setQueryDataMock);
        render(
            <QueryClientProvider client={new QueryClient()}>
                <CocktailCard cocktail={mockCocktail} mode="search" />
            </QueryClientProvider>
        );
        const addButton = screen.getByText('Add');
        fireEvent.click(addButton);
        expect(setQueryDataMock).toHaveBeenCalledWith(
            ['cocktail', 'favorites'],
            expect.any(Function)
        );
    });

    test('handles remove from cart button click', () => {
        const setQueryDataMock = jest.fn();
        // const handleRemoveFromCartToastMock = jest.fn();
        jest.spyOn(QueryClient.prototype, 'setQueryData').mockImplementation(setQueryDataMock);
        render(
            <QueryClientProvider client={new QueryClient()}>
                <CocktailCard cocktail={mockCocktail} mode="favorites" />
            </QueryClientProvider>
        );
        const removeButton = screen.getByText('Remove');
        fireEvent.click(removeButton);
        expect(setQueryDataMock).toHaveBeenCalledWith(
            ['cocktail', 'favorites'],
            expect.any(Function)
        );
        // expect(handleRemoveFromCartToastMock).toHaveBeenCalledTimes(1);
    });

    test('matches snapshot', () => {
        const { asFragment } = render(
            <QueryClientProvider client={new QueryClient()}>
                <CocktailCard cocktail={mockCocktail} mode="home" />
            </QueryClientProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});