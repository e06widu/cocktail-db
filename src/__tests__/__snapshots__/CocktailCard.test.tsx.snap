// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CocktailCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card-container"
  >
    <div>
      Mojito
    </div>
    <img
      alt="Mojito"
      class="thumb-image"
      loading="lazy"
      src="mojito.jpg"
    />
    <div>
      Cocktail
    </div>
  </div>
</DocumentFragment>
`;
